src/v/cluster/rm_stm.cc

```diff
@@ -582,49 +582,61 @@ ss::future<tx_errc> rm_stm::do_commit_tx(
-            co_return tx_errc::none;
-        }
+    model::tx_seq transaction_tx_seq;
+    if (!is_transaction_ga()) {
+        auto preparing_it = _mem_state.preparing.find(pid);
+
+        if (preparing_it != _mem_state.preparing.end()) {
+            if (preparing_it->second.tx_seq > tx_seq) {
+                // - tm_stm & rm_stm failed during prepare
+                // - during recovery tm_stm recommits its previous tx
+                // - that commit (we're here) collides with "next" failed
+                // prepare it may happen only if the commit passed => acking
+                co_return tx_errc::none;
+            }
 
-        ss::sstring msg;
-        if (preparing_it->second.tx_seq == tx_seq) {
-            msg = ssx::sformat(
-              "Prepare hasn't completed => can't commit pid:{} tx_seq:{}",
-              pid,
-              tx_seq);
-        } else {
-            msg = ssx::sformat(
-              "Commit pid:{} tx_seq:{} conflicts with preparing tx_seq:{}",
-              pid,
-              tx_seq,
-              preparing_it->second.tx_seq);
-        }
+            ss::sstring msg;
+            if (preparing_it->second.tx_seq == tx_seq) {
+                msg = ssx::sformat(
+                  "Prepare hasn't completed => can't commit pid:{} tx_seq:{}",
+                  pid,
+                  tx_seq);
+            } else {
+                msg = ssx::sformat(
+                  "Commit pid:{} tx_seq:{} conflicts with preparing tx_seq:{}",
+                  pid,
+                  tx_seq,
+                  preparing_it->second.tx_seq);
+            }
 
-        if (_recovery_policy == best_effort) {
-            vlog(_ctx_log.error, "{}", msg);
-            co_return tx_errc::request_rejected;
+            if (_recovery_policy == best_effort) {
+                vlog(_ctx_log.error, "{}", msg);
+                co_return tx_errc::request_rejected;
+            }
+            vassert(false, "{}", msg);
         }
-        vassert(false, "{}", msg);
-    }
 
-    auto prepare_it = _log_state.prepared.find(pid);
+        auto prepare_it = _log_state.prepared.find(pid);
 
-    if (prepare_it == _log_state.prepared.end()) {
-        vlog(
-          _ctx_log.trace,
-          "Can't find prepare for pid:{} => replaying already comitted commit",
-          pid);
-        co_return tx_errc::none;
+        if (prepare_it == _log_state.prepared.end()) {
+            vlog(
+              clusterlog.trace,
+              "Can't find prepare for pid:{} => replaying already comitted "
+              "commit",
+              pid);
+            co_return tx_errc::none;
+        }
+        transaction_tx_seq = prepare_it->second.tx_seq;
+    } else {
+        auto tx_seqs_it = _log_state.tx_seqs.find(pid);
+        if (tx_seqs_it == _log_state.tx_seqs.end()) {
+            vlog(clusterlog.trace, "Can't find tx_seqs for pid:{}", pid);
+            co_return tx_errc::none;
```

Comment:
add a comment that it may happen only during a duplicated commit since commit is idempotent we should ack it